Extract incremental execution

lptr opened the pull request:
Move code related to incremental change detection to the `:execution` project.

lptr commented:
@wolfs please take a look; it should already work. It's probably best to look commit-by-commit.

The execution engine now handles everything related to detecting incremental changes. It also includes actually storing the state itself after execution (the same way we do for tasks, i.e. we take into account successful execution and skip doing it if there are no output changes).

I'm planning to start dissecting the tests, so that each step can be tested on its own.

wolfs commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/TaskExecuterResult.java:
```diff
@@ -16,17 +16,41 @@
      */
     Optional<OriginMetadata> getReusedOutputOriginMetadata();
 
     TaskExecuterResult NO_REUSED_OUTPUT = new TaskExecuterResult() {
```
I think the name of this constant is a little bit strange. `NO_REUSED_OUTPUT` sounds like the task did not re-use any output. For me that means that the task executed. It also says above "An empty list means the task was not executed". This is in contradiction with the outcome "ExecutionOutcome.EXECUTED_FULLY".

I don't really understand what the executer result means :(

lptr replied:
OK, I simplified this by only providing information whether or not the task was executed incrementally, as this is the only thing we need to understand at the call site.

wolfs replied:
I still don't understand what `NO_REUSED_OUTPUT` means, I think we should use a better name here. This executer result is not that the task didn't re-use any output. Any task that executes doesn't re-use the output. This is about the task not even trying to re-use the outputs.

How about `SKIPPED_WITHOUT_OUTPUTS` or just `WITHOUT_OUTPUTS`?

lptr replied:
`WITHOUT_OUTPUTS` it is.

wolfs commented on subprojects/execution/src/main/java/org/gradle/internal/execution/ExecutionOutcome.java:
```diff
@@ -17,7 +17,25 @@
+    /**
+     * The work has been executed with no incremental change information.
+     */
+    EXECUTED_FULLY
```
How about we call this outcome `REBUILT`?

lptr replied:
How about `EXECUTED`? `REBUILT` implies a previous execution, and I don't like to mix in the verb "build" here.

wolfs replied:
I find it strange that we call the outcome `EXECUTED` and have a `rebuildReason`, `RebuildExecutionStateChanges`, etc. Your point about "rebuild" implying a previous execution applies there as well. I think we should choose one name and stick with it.

So it seems that we use "incremental" to denote the incremental execution. We do that consistently. The question is what is the opposite of "incremental". Using executed here would mean that we use nothing (absence of incremental) to denote the non-incrementalness (that would imply `IncrementalExecutionStateChanges` vs. `DefaultExecutionStateChanges`). We could also use non-incremental to specify the non-incrementalness (e.g. `EXECUTED_NON_INCREMENTALLY`, `NonIncrementalExecutionStateChanges`, etc.).

We could try to use `full` execution as an opposite of incremental as well, though I think we then need to use it consistently.

Currently, I prefer to use non-incremental.

lptr replied:
`EXECUTED_NON_INCREMENTALLY` it is.

wolfs commented on subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/ExecutionTest.groovy:
```diff
@@ -167,15 +178,6 @@ class ExecutionTest extends Specification {
-        Optional<ExecutionStateChanges> getChangesSincePreviousExecution() {
-            throw new UnsupportedOperationException()
-        }
 
```
Nit: Double newline.

lptr replied:
Removed.

wolfs commented on subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/ExecutionTest.groovy:
```diff
@@ -203,4 +205,32 @@ class ExecutionTest extends Specification {
+    def execute(Supplier<UnitOfWork> work) {
+        executionStep.execute(new IncrementalChangesContext() {
+            @Override
+            Optional<ExecutionStateChanges> getChanges() {
```
I find it pretty annoying that those things need to be overridden together each time we create a context further down the road, only to pass things through. Do you think that maybe the `rebuildReason` and the execution states should be merged into one object?

lptr replied:
The execution states each, and the rebuild reason again will be added in a separate `Step`, at least if we keep the original structure (which I intend to do at least first). I'd like to review these choices later, and improve on them once we have all the stuff we want moved over to the execution engine.

wolfs commented on subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/IncrementalExecutionTest.groovy:
```diff
@@ -57,13 +59,17 @@ import org.gradle.test.fixtures.file.TestFile
 import static org.gradle.internal.execution.ExecutionOutcome.UP_TO_DATE
+import static org.gradle.internal.execution.history.changes.OutputFileChanges.OutputHandling.DETECT_ADDED
 
+// FIXME:lptr
```
Let's not forget to fix this test before we merge the PR.

wolfs commented on subprojects/execution/src/main/java/org/gradle/internal/execution/WorkExecutor.java:
```diff
@@ -16,6 +16,6 @@
 
-public interface WorkExecutor<R extends Result> {
-    R execute(UnitOfWork work);
+public interface WorkExecutor<C extends Context, R extends Result> {
```
Can we get rid of the type parameters of `WorkExecutor`? I don't see us supporting different types in the future, so why carry around the burden of the type parameters?

lptr replied:
Yes, I think in the end we will be able to remove a bunch of type parameters. Again, I'd like to first move everything over, and then do a cleanup.

wolfs replied:
I don't quite understand why you introduced the `Context` type parameter in the first place. We can do the cleanup afterwards.

wolfs commented on subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/WorkExecutorTestFixture.java:
```diff
@@ -70,7 +71,8 @@ public void close() {
 
-    WorkExecutorTestFixture(DefaultFileSystemMirror fileSystemMirror) {
+    WorkExecutorTestFixture(DefaultFileSystemMirror fileSystemMirror,
+                            ExecutionHistoryStore executionHistoryStore) {
```
I guess the users of the test fixture need to be adjusted as well.

lptr replied:
Yep, done.

lptr commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java:
```diff
@@ -262,34 +305,9 @@ public void outputsRemovedAfterFailureToLoadFromCache() {
         private boolean hasAnyOutputFileChanges(ImmutableSortedMap<String, FileCollectionFingerprint> previous, ImmutableSortedMap<String, CurrentFileCollectionFingerprint> current) {
             return !previous.keySet().equals(current.keySet())
-                || new OutputFileChanges(previous, current).hasAnyChanges();
+                || new OutputFileChanges(previous, current, IGNORE_ADDED).hasAnyChanges();
```
Well spotted. I've simplified this a little now I think it makes more sense.

lptr replied:
I also renamed the enum values to `INCLUDE_ADDED` and `EXCLUDE_ADDED`, and improved their Javadocs.

wolfs replied:
I saw you removed the enums again.

lptr commented:
@wolfs I addressed your comments, please take another look. Tomorrow I'll work on the tests. Perhaps we can merge this already if you're happy, so you can rebase the incremental transforms stuff on top of this.

wolfs commented on subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ResolveChangesStep.java:
```diff
@@ -0,0 +1,116 @@
+
+    @Override
+    public R execute(IncrementalContext context) {
+        final UnitOfWork work = context.getWork();
```
No need to use final here, right?

lptr replied:
Removed.

wolfs commented on subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java:
```diff
@@ -63,10 +58,17 @@
     Optional<? extends Iterable<String>> getChangingOutputs();
 
+    /**
+     * Whether files added are to be considered part of the output of the work.
```
I think it is not about whether files added to the output locations are considered an output of the work. When the task adds new files in an output directory, they are always considered an output of the work.

It is about whether the task is considered up-to-date if something else adds outputs to the output locations, without changing the recorded outputs of the last execution of the work. So it is about whether to allow overlapping outputs or not, right?

lptr replied:
Clarified this a bit.

wolfs commented:
@lptr I added some more comments and replied to yours. I think we can merge this PR (as long as tests are passing) and address the  remaining issues as a follow up PR.
